docs/rfcs/021-raii-runtime.md

```diff
@@ -0,0 +1,229 @@
+
+The interface:
+
+```rust
+impl TenantAccessor {
+    /// Creates a new accessor. Check the state is not `TenantState::Infant`.
+    /// Spawns a closure in tokio runtime to wait for the state to become dead.
+    /// Then, the tenant is shut down.
+    pub fn new(tenant: Tenant) -> Self;
+
+    /// Subscribe for the tenant drop. If the tenant is already dropped, then
+    /// watcher will also reflect that.
+    pub fn subscribe_for_shutdown(&self) -> DropWatcher;
+
+    /// Applies the specified function to the tenant. Returns error if the tenant is
+    /// not ready to launch the closure.
+    pub fn with_tenant<F, O>(&self, func: F) -> anyhow::Result<O>
+    where
+    F: FnOnce(TenantSyncRef) -> O {}
+
+    /// Applies the specified function to the tenant. Returns error if the tenant is
+    /// not ready to launch the closure.
+    pub async fn with_tenant_async<F, T, O>(&self, func: F) -> anyhow::Result<O>
+    where
+    F: FnOnce(TenantAsyncRef) -> T,
+    T: Future<Output = O> {}
+}
+
+impl Drop for TenantAccessor {
+    // schedule a drop of itself in tokio runtime with necessary joins...
+}
+```
+
+### `TenantAccessor` and states
+
+This is how the accessor behaves when it sees different states:
+
+1. `Infant` - we cannot create a guard around the tenant at this stage. We should start loading some layers and become `Loading` or `Downloading` first and only after creating an accessor.
+2. `Loading` - we cannot run any operations on the tenant until it is either in an `Active` or `Broken` state.
+3. `Broken` - we cannot run any operations on the tenant.
+4. `Downloading` - we cannot run any operations on the tenant until it is either in an `Active` or `Broken` state.
+5. `Active` - the tenant can run any closures. **That means it's the only way to run any closures on tenants**.
+6. `ShuttingDown` - we cannot run any operations on the tenant.
+7. `Shutdown` - we cannot run any operations on the tenant.
+
+### `TenantAccessor` and cancelations
+
+One great question is: what guarantees do we have about tenants inside the closures we're running?
+
+After we introduced the `with_tenant` functions, there are a bunch of closures running. The only way to run a closure is when we are under an `Active` state. What if one closure wants to run under this state, and another decides to shut down the tenant? Do we want to continue executing?
+
+In this specific RFC:
+
+1. We use a `StateKeeper` / `StateSubscriber` / `StateHider` primitive to ensure all closures are gone, and running closures can't tell the difference.
+2. It's proposed to run only non-cancellable short-living closures and check in the code reviews that the closure won't run for a long time. This should be easy to spot. Moreover: _all of the current code does not use long-running functions inside tenants_, even when GC or compaction is run.
```

Comment:
Workaround from thread per core world. When your code runs for more than 500ms without releasing control back runtime will emit a backtrace.

Is gc expected to run in such closure? Flushing a layer during compaction can take significant time.

In general you cannot be sure. There are context switches, you can end up waiting for some lock for a long time. Looks like a termination problem to me. For sure it is not easy to spot.

Reply:
Yes, it's expected. I guess it's better to require only sync closures to run short, and async closures are expected to progress with no deadlocks, or instead, we'll have a never-dying tenant.